control-plane/connect-inject/controllers/pod/pod_controller.go

```diff
@@ -110,7 +120,13 @@ func (r *Controller) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
 	// we need to remove the Workload from Consul.
 	if k8serrors.IsNotFound(err) {
 
-		if err := r.deleteWorkload(ctx, req.NamespacedName); err != nil {
+		if err := r.deleteHealthStatus(ctx, req.NamespacedName); err != nil {
+			errs = multierror.Append(errs, err)
+		}
+
+		// TODO: clean up ACL Tokens
+
+		if err := r.deleteProxyConfiguration(ctx, req.NamespacedName); err != nil {
```

Comment:
Does order matter at all here on the Consul side before the workload is deleted (i.e. should we fast-fail at any point)?

Also wondering whether we can rely on Consul controllers to do these deletes for us - I'd expect clients to be able to rely on Consul to manage cascading deletes of dependent resources via `Owner` w/o managing that part of the lifecycle explicitly. Is that something we're not planning on ATM? (I can't find any references in the RFCs that describe owner links, but it seems implied by the k8s-oriented data model)

Reply:
Yeah, IMHO, order definitely matters. I was trying to remove the health check first to prevent routing to the workload. ProxyConfiguration should go last for deletion. Of course, since everything is eventually consistent, perhaps it's less important than I think.

That could be where Consul comes in. I have observed that Consul does handle cascading deletes. I kept the explicit deletions to be safe, but it might help with the consistency to let Consul do it. I will update it so HealthStatus is deleted implicitly.  

Thinking about the eventual consistency, though, maybe we need to wait until the workload is removed to delete the proxy configuration, lest we change the behavior before the Consul reconcile completes?

Reply:
That’s a good point about ProxyConfiguration. It makes me feel like we should have that point to a Workload as owner too, so there’s a cascading deletion relationship, and perhaps validate on the server side that selector and owner are 1-1 when owner is specified. IMO it’s not super robust to rely on this controller to manage that lifecycle since ProxyConfiguration could easily be orphaned if the pod goes away when the controller is erroring or otherwise MIA. Do you think that makes sense as a question to pose or is the lack of guaranteed GC a constraint we’ve chosen to live with explicitly somewhere?

Reply:
It's a good question to ask. It seems like if we start by deleting the owner, then deleting the orphans in GC, that would be sufficient for what we need. I think the problem with using an Owner ref is that it would be optional. I'm not sure either way if the controller arch supports it.